### PR TITLE
sql: add setting to clone parent partitions for deinterleaving

### LINKED ISSUE #66193: Support inheriting partitions from Parent Table after migration of child table from  INTERLEAVE
**Is your feature request related to a problem? Please describe.**
Current Convert interleaved tables does not support passing parent partition setting to child tables.

**Describe the solution you'd like**
ALTER PRIMARY KEY support partition configuration from parent table or ability to partition a child table prior to migration.

**Describe alternatives you've considered**
add partition after running ALTER PRIMARY KEY 

Example Schema

https://drive.google.com/file/d/1sm4Df9Iok7S84S_XSmJijngefz1t_buU/view?usp=sharing

Roachprod with 3 nodes ( east1, west2, central1)

restore defaultdb [BACKUP DATABASE defaultdb TO 'nodelocal://1/interleaved']
restore table defaultdb.public.* from 'nodelocal://1/interleaved' WITH into_db = 'newdb';
        job_id       |  status   | fraction_completed |  rows   | index_entries |   bytes
---------------------+-----------+--------------------+---------+---------------+-------------
  676222251025825793 | succeeded |                  1 | 8087601 |             0 | 1112658935
(1 row)

Time: 9.935s total (execution 9.935s / network 0.000s)

```CREATE TABLE public.session (
  session_id STRING(100) NOT NULL,
  affliate STRING(50) NOT NULL,
  channel STRING(50) NOT NULL,
  language STRING(20) NOT NULL,
  date_creation TIMESTAMP NOT NULL,
  date_updated TIMESTAMP NOT NULL,
  status STRING(20) NOT NULL,
  platform STRING(50) NULL,
  query_id STRING(100) NULL,
  CONSTRAINT "primary" PRIMARY KEY (session_id ASC),
  FAMILY "primary" (session_id, affliate, channel, language, date_creation, date_updated, status, platform, query_id)
) PARTITION BY RANGE (session_id) (
  PARTITION east VALUES FROM ('E-') TO ('F-'), 
  PARTITION west VALUES FROM ('W-') TO ('X-'), 
  PARTITION central VALUES FROM ('C-') TO ('D-')
);
ALTER PARTITION central OF INDEX defaultdb.public.session@primary CONFIGURE ZONE USING
  range_min_bytes = 1048576,
  range_max_bytes = 67108864,
  gc.ttlseconds = 90000,
  num_replicas = 3,
  constraints = '[]',
  lease_preferences = '[[+region=us-central1]]';
ALTER PARTITION east OF INDEX defaultdb.public.session@primary CONFIGURE ZONE USING
  range_min_bytes = 1048576,
  range_max_bytes = 67108864,
  gc.ttlseconds = 90000,
  num_replicas = 3,
  constraints = '[]',
  lease_preferences = '[[+region=us-east1]]';
ALTER PARTITION west OF INDEX defaultdb.public.session@primary CONFIGURE ZONE USING
  range_min_bytes = 1048576,
  range_max_bytes = 67108864,
  gc.ttlseconds = 90000,
  num_replicas = 3,
  constraints = '[]',
  lease_preferences = '[[+region=us-west2]]'

CREATE TABLE public.device_identifier (
  id STRING(100) NOT NULL,
  session_id STRING(100) NOT NULL,
  device_id STRING(50) NULL,
  device_name STRING(50) NULL,
  device_make STRING(50) NULL,
  device_macaddress STRING(50) NULL,
  device_model STRING(50) NULL,
  device_serialno STRING(50) NULL,
  date_creation TIMESTAMP NOT NULL,
  date_updated TIMESTAMP NOT NULL,
  CONSTRAINT "primary" PRIMARY KEY (session_id ASC, id ASC),
  CONSTRAINT fk_session_id_ref_session FOREIGN KEY (session_id) REFERENCES public.session(session_id) ON DELETE CASCADE,
  FAMILY "primary" (id, session_id, device_id, device_name, device_make, device_macaddress, device_model, device_serialno, date_creation, date_updated)
) INTERLEAVE IN PARENT public.session (session_id)

CREATE TABLE public.customer_identifier (
  session_id STRING(100) NOT NULL,
  customer_key STRING(50) NOT NULL,
  customer_value STRING(50) NOT NULL,
  date_creation TIMESTAMP NOT NULL,
  date_updated TIMESTAMP NOT NULL,
  CONSTRAINT "primary" PRIMARY KEY (session_id ASC, customer_key ASC),
  CONSTRAINT fk_session_id_ref_session FOREIGN KEY (session_id) REFERENCES public.session(session_id) ON DELETE CASCADE,
  FAMILY "primary" (session_id, customer_key, customer_value, date_creation, date_updated)
) INTERLEAVE IN PARENT public.session (session_id)

```

image

migrate  device_identifier

```

alter table device_identifier ALTER PRIMARY KEY USING COLUMNS (session_id, id);


NOTICE: primary key changes are finalized asynchronously; further schema changes on this table may be restricted until the job completes
ALTER TABLE

Time: 47.347s total (execution 1.078s / network 46.270s)

 show create device_identifier;
     table_name     |                                                                     create_statement
--------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------
  device_identifier | CREATE TABLE public.device_identifier (
                    |     id STRING(100) NOT NULL,
                    |     session_id STRING(100) NOT NULL,
                    |     device_id STRING(50) NULL,
                    |     device_name STRING(50) NULL,
                    |     device_make STRING(50) NULL,
                    |     device_macaddress STRING(50) NULL,
                    |     device_model STRING(50) NULL,
                    |     device_serialno STRING(50) NULL,
                    |     date_creation TIMESTAMP NOT NULL,
                    |     date_updated TIMESTAMP NOT NULL,
                    |     CONSTRAINT "primary" PRIMARY KEY (session_id ASC, id ASC),
                    |     CONSTRAINT fk_session_id_ref_session FOREIGN KEY (session_id) REFERENCES public.session(session_id) ON DELETE CASCADE,
                    |     UNIQUE INDEX device_identifier_session_id_id_key (session_id ASC, id ASC),
                    |     FAMILY "primary" (id, session_id, device_id, device_name, device_make, device_macaddress, device_model, device_serialno, date_creation, date_updated)
                    | )
(1 row)

Time: 1.059s total (execution 1.058s / network 0.000s)


```

migrate  customer_identifier

```
alter table customer_identifier ALTER PRIMARY KEY USING COLUMNS (session_id,customer_key);
NOTICE: primary key changes are finalized asynchronously; further schema changes on this table may be restricted until the job completes
ALTER TABLE

Time: 33.536s total (execution 1.022s / network 32.514s)

 show create customer_identifier;
      table_name      |                                                      create_statement
----------------------+-----------------------------------------------------------------------------------------------------------------------------
  customer_identifier | CREATE TABLE public.customer_identifier (
                      |     session_id STRING(100) NOT NULL,
                      |     customer_key STRING(50) NOT NULL,
                      |     customer_value STRING(50) NOT NULL,
                      |     date_creation TIMESTAMP NOT NULL,
                      |     date_updated TIMESTAMP NOT NULL,
                      |     CONSTRAINT "primary" PRIMARY KEY (session_id ASC, customer_key ASC),
                      |     CONSTRAINT fk_session_id_ref_session FOREIGN KEY (session_id) REFERENCES public.session(session_id) ON DELETE CASCADE,
                      |     UNIQUE INDEX customer_identifier_session_id_customer_key_key (session_id ASC, customer_key ASC),
                      |     FAMILY "primary" (session_id, customer_key, customer_value, date_creation, date_updated)


```

Both child tables lack the partition config or parent and lease are not correct as a result

image

image

Notice after partitioning the child tables, a UNIQUE INDEX is created after migration which either has to partitioned or removed

```
 alter table device_identifier PARTITION BY RANGE (session_id) (PARTITION east VALUES FROM ('E-') TO ('F-'),PARTITION west VALUES FROM ('W-') TO ('X-'),PARTITION central VALUES FROM ('C-') TO ('D-'));
ALTER TABLE

Time: 3.062s total (execution 0.955s / network 2.107s)

 alter table customer_identifier PARTITION BY RANGE (session_id) (PARTITION east VALUES FROM ('E-') TO ('F-'),PARTITION west VALUES FROM ('W-') TO ('X-'),PARTITION central VALUES FROM ('C-') TO ('D-'));
ALTER TABLE

Time: 2.946s total (execution 0.956s / network 1.990s)

show create customer_identifier;
      table_name      |                                                      create_statement
----------------------+-----------------------------------------------------------------------------------------------------------------------------
  customer_identifier | CREATE TABLE public.customer_identifier (
                      |     session_id STRING(100) NOT NULL,
                      |     customer_key STRING(50) NOT NULL,
                      |     customer_value STRING(50) NOT NULL,
                      |     date_creation TIMESTAMP NOT NULL,
                      |     date_updated TIMESTAMP NOT NULL,
                      |     CONSTRAINT "primary" PRIMARY KEY (session_id ASC, customer_key ASC),
                      |     CONSTRAINT fk_session_id_ref_session FOREIGN KEY (session_id) REFERENCES public.session(session_id) ON DELETE CASCADE,
                      |     UNIQUE INDEX customer_identifier_session_id_customer_key_key (session_id ASC, customer_key ASC),
                      |     FAMILY "primary" (session_id, customer_key, customer_value, date_creation, date_updated)
                      | ) PARTITION BY RANGE (session_id) (
                      |     PARTITION east VALUES FROM ('E-') TO ('F-'),
                      |     PARTITION west VALUES FROM ('W-') TO ('X-'),
                      |     PARTITION central VALUES FROM ('C-') TO ('D-')
                      | )
                      | -- Warning: Partitioned table with no zone configurations.
(1 row)

Time: 1.057s total (execution 1.057s / network 0.000s)

show create device_identifier;
     table_name     |                                                                     create_statement
--------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------
  device_identifier | CREATE TABLE public.device_identifier (
                    |     id STRING(100) NOT NULL,
                    |     session_id STRING(100) NOT NULL,
                    |     device_id STRING(50) NULL,
                    |     device_name STRING(50) NULL,
                    |     device_make STRING(50) NULL,
                    |     device_macaddress STRING(50) NULL,
                    |     device_model STRING(50) NULL,
                    |     device_serialno STRING(50) NULL,
                    |     date_creation TIMESTAMP NOT NULL,
                    |     date_updated TIMESTAMP NOT NULL,
                    |     CONSTRAINT "primary" PRIMARY KEY (session_id ASC, id ASC),
                    |     CONSTRAINT fk_session_id_ref_session FOREIGN KEY (session_id) REFERENCES public.session(session_id) ON DELETE CASCADE,
                    |     UNIQUE INDEX device_identifier_session_id_id_key (session_id ASC, id ASC),
                    |     FAMILY "primary" (id, session_id, device_id, device_name, device_make, device_macaddress, device_model, device_serialno, date_creation, date_updated)
                    | ) PARTITION BY RANGE (session_id) (
                    |     PARTITION east VALUES FROM ('E-') TO ('F-'),
                    |     PARTITION west VALUES FROM ('W-') TO ('X-'),
                    |     PARTITION central VALUES FROM ('C-') TO ('D-')
                    | )
                    | -- Warning: Partitioned table with no zone configurations.
(1 row)

Time: 1.060s total (execution 1.060s / network 0.000s)
```

The alternative of partition a child fails

```
alter table query_identifier PARTITION BY RANGE (session_id) (PARTITION east VALUES FROM ('E-') TO ('F-'),PARTITION west VALUES FROM ('W-') TO ('X-'),PARTITION central VALUES FROM ('C-') TO ('D-'));
ERROR: relation "query_identifier" (61): cannot set a zone config for interleaved index primary; set it on the root of the interleaved hierarchy instead

```

perhaps the added  UNIQUE INDEX is a result of the constraint.

### PR DESCRIPTION
**Pull request by ajwerner**
See the release note.

Fixes #66193.

Release note (sql change): Added a new session variable
`enable_copying_partitioning_when_deinterleaving_table` (defaulted to a new
cluster setting
`sql.defaults.copy_partitioning_when_deinterleaving_table.enabled`) which will
change the behavior of ALTER PRIMARY KEY when performing a change which retains
the same primary key but removes an INTERLEAVE INTO clause. When this variable
is true and such an ALTER PRIMARY KEY is run, the partitioning and zone
configuration which applied to the root of the interleave will not be applied
to the new primary index.

### CODE REVIEW

**Comment by ajwerner** [comment]
cc @vy-ton, @drewdeally

File: pkg/sql/vars.go [hunk 1]
```diff
@@ -1382,6 +1382,24 @@ var varGen = map[string]sessionVar{
 			return experimentalComputedColumnRewrites.Get(sv)
 		},
 	},
+
+	`enable_copying_partitioning_when_deinterleaving_table`: {
+		GetStringVal: makePostgresBoolGetStringValFn(`enable_experimental_stream_replication`),
```
Comment (otan) [inline pkg/sql/vars.go:1387]: nit: use constant?

File: pkg/ccl/logictestccl/testdata/logic_test/partitioning_deinterleave [hunk 2]
```diff
@@ -0,0 +1,489 @@
...
+  k INT8 NOT NULL,
+  CONSTRAINT "primary" PRIMARY KEY (i ASC),
+  UNIQUE INDEX grandchild_i_j_k_key (i ASC, j ASC, k ASC),
+  INDEX idx_k_j_i (k ASC, j ASC, i ASC) PARTITION BY LIST (k) (
+    PARTITION negative VALUES IN ((DEFAULT), (-1)),
+    PARTITION zero VALUES IN ((0)),
+    PARTITION positive VALUES IN ((1))
+  ),
+  FAMILY "primary" (i, j, k)
+);
```
Comment (otan) [inline pkg/ccl/logictestccl/testdata/logic_test/partitioning_deinterleave:236]: am i missing something or is a PK partitioning missing here?

Comment (ajwerner) [reply pkg/ccl/logictestccl/testdata/logic_test/partitioning_deinterleave:250]: 🤦 fixed.

File: pkg/sql/vars.go (hunk 1 above)
Comment (ajwerner) [reply pkg/sql/vars.go:1387]: The entire rest of the file uses the string twice 🤷

**Review by otan — APPROVED** [review]
hehe i thought i was going crazy

**Comment by ajwerner** [comment]
TFTR

bors r=otan

**Pull request merged**